### PR TITLE
 admin/self-hosted-etcd.md: add documentation on self-hosted etcd on Tectonic

### PR DESCRIPTION
**Pull request by radhikapc**
ref: https://github.com/coreos/tectonic-installer/issues/1383

### CODE REVIEW

File: Documentation/admin/configure-selfhosted-etcd.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
+# Self-hosted etcd cluster on Tectonic
```
Comment (xiang90) [inline Documentation/admin/configure-selfhosted-etcd.md]: probably change the doc file name to just self-hosted-ecd.md?

this doc is more generic than configuring.

File: Documentation/admin/configure-selfhosted-etcd.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+* etcd can automatically recover from machine failures
+* etcd can run on the same host as the control plane
+
+## Operating self-hosted etcd
+
+Tectonic uses etcd operator, which runs on the same Kubernetes cluster as a pod, to control a self-hosted etcd. The etcd operator handles bootstrapping, quorum maintenance, reconfiguring cluster membership, creating backups, handling disaster recovery, and monitoring critical events related to a self-hosted etcd cluster.
+
+### Enabling self-hosted etcd in Tectonic
+
+To launch and configure a self-hosted etcd cluster and an etcd operator in Tectonic, select the *Create self-hosted etcd cluster(alpha)* checkbox. This option is found under *etcd Nodes* in the *Cluster Info* page of Tectonic Installer.
```
Comment (xiang90) [inline Documentation/admin/configure-selfhosted-etcd.md]: in which step can we select the checkbox? ideally we can have a graph to show that?

File: Documentation/admin/configure-selfhosted-etcd.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+### Backing up self-hosted etcd
+
+Automatic backups can be configured for disaster recovery. Similar to the `version` field, add a `backup` policy field to the etcd resource spec on Tectonic Console. Under *Operator*, click *etcd Clusters*, and select the etcd cluster. Click *YAML* to view or edit the resource spec. A sample backup spec is given below:
+
+```YAML
+  backup:
+    backupIntervalInSecond: 30
+    maxBackups: 5
+    storageType: PersistentVolume
```
Comment (xiang90) [inline Documentation/admin/configure-selfhosted-etcd.md]: change to s3?

File: Documentation/admin/configure-selfhosted-etcd.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    pv:
+      volumeSizeInMB: 512
+```    
+
+The etcd operator will create a backup pod when backup policy is configured. The process periodically takes backups from the etcd cluster, and upload the backups to the storage defined in the backup policy. Backups can be saved as Persistent Volume (PV) on GCE or AWS, or as S3 bucket on AWS.
+
+#### Persistent volume
```
Comment (xiang90) [inline Documentation/admin/configure-selfhosted-etcd.md]: we can remove PV. we suggest users to use s3 for tectonic.

File: Documentation/admin/about-selfhosted-etcd.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+## About self-hosted etcd
+
+A self-hosted etcd is an auto-managed cluster of pods running on Kubernetes. It acts as the primary data store for the same Kubernetes cluster that it is running on. A self-hosted etcd operates significantly different from an external or a provisioned etcd. With self-hosted etcd enabled, Tectonic handles installation, upgrade, backup, and disaster recovery of a self-hosted etcd.
+
+The objective of having a self-hosted etcd cluster is to run all the Kubernetes master components, including etcd, of a Tectonic cluster on Kubernetes itself (self-hosted Kubernetes).
+
+This setup provides the following benefits:
+
+* etcd is TLS secured
+* etcd is HA-enabled
```
Comment (robszumski) [inline Documentation/admin/about-selfhosted-etcd.md]: etcd is **highly available**

File: Documentation/admin/about-selfhosted-etcd.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+* etcd can be automatically backed up to a stable storage
+* etcd is upgradable
+* etcd can automatically recover from machine failures
+* etcd can run on the same host as the control plane
+
+## Operating self-hosted etcd
+
+Tectonic uses etcd operator, which runs on the same Kubernetes cluster as a pod, to control a self-hosted etcd. The etcd operator handles bootstrapping, quorum maintenance, reconfiguring cluster membership, creating backups, handling disaster recovery, and monitoring critical events related to a self-hosted etcd cluster.
```
Comment (robszumski) [inline Documentation/admin/about-selfhosted-etcd.md]: link to either https://coreos.com/operators/ or https://github.com/coreos/etcd-operator?

File: Documentation/admin/about-selfhosted-etcd.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    <a href="../img/self-hosted-etcd.png" class="co-m-screenshot">
+      <img src="../img/self-hosted-etcd.png">
+      <div class="co-m-screenshot-caption">Enable self-hosted etcd</div>
+    </a>
+  </div>
+</div>
+
+### Installation workflow
+
+Tectonic uses `bootkube` to install a self-hosted Kubernetes cluster with self-hosted etcd enabled. `bootkube` creates an external boot etcd cluster member to initiate the installation. Boot etcd cluster member refers to the initial etcd member created externally as a local static pod during the bootstrap phase. The complete installation workflow is summarized as follows:
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: "Tectonic uses [`bootkube`][LINK] to install a Kubernetes cluster with a self-hosted etcd service. `Bootkube` creates an external etcd cluster member for bootstrapping the installation. This initial etcd cluster member is crated and only used during initial setup.

The following list summarizes the self-hosted etcd bootstrap and installation process:"

File: Documentation/admin/about-selfhosted-etcd.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+1. Starts a temporary external single-member etcd cluster
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: 1. Start a temporary, external etcd cluster with a single member

File: Documentation/admin/about-selfhosted-etcd.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+2. Starts a temporary Kubernetes cluster using the temporary etcd cluster’s address
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Starts/Start/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+3. Starts an etcd operator in the temporary Kubernetes cluster
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Starts/Start/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+4. Adds one of the Kubernetes member pods into the etcd cluster by using the etcd operator
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: 4. Use the etcd operator to add one of the Kubernetes etcd pods as a cluster member

File: Documentation/admin/about-selfhosted-etcd.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+5. Creates an [etcd service][etcd-service] for the etcd cluster
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Creates/Create/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+6. Starts a self-hosted Kubernetes cluster using the etcd service’s cluster address
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Starts/Start/
s/a/the permanent/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+7. Removes the temporary external single-member etcd cluster
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Removes/Remove/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+8. Stops the temporary Kubernetes cluster
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/Stops/Stop/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+## Maintaining self-hosted etcd
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/transient//

File: Documentation/admin/about-selfhosted-etcd.md [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+The etcd operator is responsible for the maintenance operations, such as upgrade, back up, and recovery from transient failures.
+
+### Upgrading self-hosted etcd
+
+Upgrading a self-hosted etcd is easy and automatic. A self-hosted etcd will have a corresponding [Customized Resource Definition (CRD)][CRD] item in the `kube-system` namespace. To upgrade manually, specify the expected etcd `version` in the YAML editor available on Tectonic Console.
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: first sentence: replace with: "Self-hosted etcd clusters can be automatically upgraded by the etcd operator."

Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/will have/has/
s/item/object/
s/expected/desired/
s/on/in/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+Under *Operator*, click *etcd Clusters*, and select the desired etcd. Click *YAML* to edit the resource spec. Click *Save* for etcd operator to automatically upgrade the etcd by updating the pod image.
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: Operators?
s/for etcd operator/for the etcd operator/
s/the etcd/etcd/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+An example resource specification for a self-hosted etcd is given below:
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/resource specification/manifest/ (?)

File: Documentation/admin/about-selfhosted-etcd.md [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+  version: 3.1.4
+  backup:
+    backupIntervalInSecond: 30
+    maxBackups: 5
+    storageType: S3
+    pv:
+      volumeSizeInMB: 512
+```
+
+A Tectonic upgrade includes an etcd version update. When an update exists, Tectonic updates the `version` field of the CRD item, and trigger an etcd update through etcd operator.
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: "Most Tectonic updates include an etcd version update. When such an update exists, Tectonic adjusts the `version` field in the etcd CRD to trigger the etcd operator to update to that version."

File: Documentation/admin/about-selfhosted-etcd.md [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+### Backing up self-hosted etcd
+
+Automatic backups can be configured for disaster recovery. Similar to the `version` field, add a `backup` policy field to the etcd resource spec on Tectonic Console. Under *Operator*, click *etcd Clusters*, and select the etcd cluster. Click *YAML* to view or edit the resource spec. A sample backup spec is given below:
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: Operator_s_?

File: Documentation/admin/about-selfhosted-etcd.md [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+  backup:
+    backupIntervalInSecond: 1800
+    maxBackups: 5
+    storageType: "S3"
+    s3:
+      s3Bucket: <S3-bucket-name>
+      awsSecret: <aws-secret-name>
+```
+
+When the backup policy is configured, the etcd operator creates a backup pod. The process periodically takes backups from the etcd cluster, and upload the backups to the storage defined in the backup policy, in this example, S3. Backups can be saved as S3 bucket on AWS, or as Persistent Volume (PV) on GCE or AWS. However, Tectonic recommends using S3 for storing etcd backups.
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/the/a/
s/upload/writes/
s/policy,/policy –/
s/S3/Amazon S3/
s/as S3/as an S3/
s/as Persistent/as a Persistent/
s/However,//

File: Documentation/admin/about-selfhosted-etcd.md [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+For information on taking etcd backup, see [Storing etcd backup][backup].
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/taking/creating/
s/backup/backups/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+
+### Recovering self-hosted etcd from a transient failure
+
+The etcd operator monitors the health of a self-hosted etcd by querying the pod states of each member. If a pod is failed, the etcd operator deletes the corresponding member from the self-hosted etcd. It replaces the failed one with a new pod, and add the new member to etcd. This works well for handling permanent node failures. When a node goes offline for maintenance or upgrades, etcd operator attempts to move the etcd pod running on that node to another one.
```
Comment (joshix) [inline Documentation/admin/about-selfhosted-etcd.md]: s/is failed/has failed/
s/and add the/and adds the/
s/etcd operator attempts/the etcd operator attempts/

File: Documentation/admin/about-selfhosted-etcd.md [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+
+1. Starts a temporary external single-member etcd cluster
+2. Starts a temporary Kubernetes cluster using the temporary etcd cluster’s address
+3. Starts an etcd operator in the temporary Kubernetes cluster
+4. Adds one of the Kubernetes member pods into the etcd cluster by using the etcd operator
+5. Creates an [etcd service][etcd-service] for the etcd cluster
+6. Starts a self-hosted Kubernetes cluster using the etcd service’s cluster address
+7. Removes the temporary external single-member etcd cluster
+8. Stops the temporary Kubernetes cluster
+
```
Comment (robszumski) [inline Documentation/admin/about-selfhosted-etcd.md]: maybe add a comment re-affirming that all of this is automatic via the installer

File: Documentation/admin/about-selfhosted-etcd.md [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+## Maintaining self-hosted etcd
+
+The etcd operator is responsible for the maintenance operations, such as upgrade, back up, and recovery from transient failures.
+
+### Upgrading self-hosted etcd
```
Comment (robszumski) [inline Documentation/admin/about-selfhosted-etcd.md]: @xiang90 no one should do this manually right? should this be removed?

File: Documentation/admin/about-selfhosted-etcd.md [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+
+### Recovering self-hosted etcd from a transient failure
+
+The etcd operator monitors the health of a self-hosted etcd by querying the pod states of each member. If a pod is failed, the etcd operator deletes the corresponding member from the self-hosted etcd. It replaces the failed one with a new pod, and add the new member to etcd. This works well for handling permanent node failures. When a node goes offline for maintenance or upgrades, etcd operator attempts to move the etcd pod running on that node to another one.
+
+When a node is partitioned from the API server, the pod status is changed from running to failed by the Kubernetes controller manager. As mentioned above, the etcd operator removes the etcd member on the partitioned node, as the pod status becomes failed. However, the etcd pod, which is deleted from self-hosted etcd, continues to run on the partitioned node until it reconnects to the Kubernetes API server. The self-hosted etcd handles this change gracefully.
+
+The self-hosted etcd cannot recover from failures when the Kubernetes control plane is unavailable, because it relies on Kubernetes to check the pod health and schedule new pods.
```
Comment (robszumski) [inline Documentation/admin/about-selfhosted-etcd.md]: ..but you can recover via...

File: Documentation/admin/about-selfhosted-etcd.md (hunk 26 above)
Comment (xiang90) [inline Documentation/admin/about-selfhosted-etcd.md]: this explains how etcd operator upgrades etcd at high level. this doc explains how self hosted etcd works in gneral.

**Comment by joshix** [comment]
last but not least, recommend file name `self-hosted-etcd.md`

File: Documentation/admin/about-selfhosted-etcd.md (hunk 16 above)
Comment (radhikapc) [inline Documentation/admin/about-selfhosted-etcd.md]: maintenance include both stable (eg: upgrade) and transient.

File: Documentation/admin/about-selfhosted-etcd.md (hunk 23 above)
Comment (radhikapc) [inline Documentation/admin/about-selfhosted-etcd.md]: `taking backup` is regular usage, like taking a snapshot

File: Documentation/admin/self-hosted-etcd.md [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+1. Start a temporary external single-member etcd cluster
+2. Start a temporary Kubernetes cluster using the temporary etcd cluster’s address
+3. Start an etcd operator in the temporary Kubernetes cluster
+4. Use the etcd operator to add one of the Kubernetes etcd pods as a cluster member
+5. Create an [etcd service][etcd-service] for the etcd cluster
+6. Start the permanent self-hosted Kubernetes cluster using the etcd service’s cluster address
+7. Remove the temporary external single-member etcd cluster
+8. Stop the temporary Kubernetes cluster
+
+In Tectonic, The etcd operator perform these bootstrapping and installation process.
```
Comment (joshix) [inline Documentation/admin/self-hosted-etcd.md]: s/these bootstrapping and installation process/this installation process/

File: Documentation/admin/self-hosted-etcd.md [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+
+## Maintaining self-hosted etcd
+
+The etcd operator is responsible for the maintenance operations, such as upgrade, back up, and recovery from transient failures.
+
+### Upgrading self-hosted etcd
+
+Self-hosted etcd clusters can be automatically upgraded by the etcd operator. A self-hosted etcd has a corresponding [Customized Resource Definition (CRD)][CRD] object in the `kube-system` namespace. To upgrade manually, specify the desired etcd `version` in the YAML editor available in Tectonic Console.
```
Comment (joshix) [inline Documentation/admin/self-hosted-etcd.md]: s/Customized/Custom/

File: Documentation/admin/self-hosted-etcd.md [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+    maxBackups: 5
+    storageType: "S3"
+    s3:
+      s3Bucket: <S3-bucket-name>
+      awsSecret: <aws-secret-name>
+```
+
+When a backup policy is configured, the etcd operator creates a backup pod. The process periodically takes backups from the etcd cluster, and writes the backups to the storage defined in the backup policy, in this example, Amazon S3. Backups can be saved as an S3 bucket on AWS, or as a Persistent Volume (PV) on GCE or AWS. Tectonic recommends using S3 for storing etcd backups.
+
+For information on creating etcd backup, see [Storing etcd backups][backup].
```
Comment (joshix) [inline Documentation/admin/self-hosted-etcd.md]: s/creating etcd backup/creating etcd backups/

File: Documentation/admin/self-hosted-etcd.md [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+
+The etcd operator monitors the health of a self-hosted etcd by querying the pod states of each member. If a pod has failed, the etcd operator deletes the corresponding member from the self-hosted etcd. It replaces the failed one with a new pod, and adds the new member to etcd. This works well for handling permanent node failures. When a node goes offline for maintenance or upgrades, the etcd operator attempts to move the etcd pod running on that node to another one.
+
+When a node is partitioned from the API server, the pod status is changed from running to failed by the Kubernetes controller manager. As mentioned above, the etcd operator removes the etcd member on the partitioned node, as the pod status becomes failed. However, the etcd pod, which is deleted from self-hosted etcd, continues to run on the partitioned node until it reconnects to the Kubernetes API server. The self-hosted etcd handles this change gracefully.
+
+The self-hosted etcd cannot recover from failures when the Kubernetes control plane is unavailable, because it relies on Kubernetes to check the pod health and schedule new pods. Ensure that you create backup periodically to prevent data loss.
```
Comment (joshix) [inline Documentation/admin/self-hosted-etcd.md]: s/create backup/create backups/

File: Documentation/admin/self-hosted-etcd.md [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+1. Start a temporary external single-member etcd cluster
+2. Start a temporary Kubernetes cluster using the temporary etcd cluster’s address
+3. Start an etcd operator in the temporary Kubernetes cluster
+4. Use the etcd operator to add one of the Kubernetes etcd pods as a cluster member
+5. Create an [etcd service][etcd-service] for the etcd cluster
+6. Start the permanent self-hosted Kubernetes cluster using the etcd service’s cluster address
+7. Remove the temporary external single-member etcd cluster
+8. Stop the temporary Kubernetes cluster
+
+In Tectonic, The etcd operator perform this installation process.
```
Comment (joshix) [inline Documentation/admin/self-hosted-etcd.md]: s/The/the/
s/perform/performs/

**Comment by robszumski** [comment]
Just docs, merging

**Pull request merged**